### PR TITLE
Pete macaluso

### PR DESCRIPTION
**Pull request by pmacaluso3**

### CODE REVIEW

File: arrowhead/arrowhead.rb [hunk 1]
```diff
@@ -15,20 +14,14 @@ class Arrowhead
...
-      else
-        raise "Unknown shape value. Are you sure you know what you're talking about?"
-      end
-    else
-      raise "Unknown region, please provide a valid region."
+  def self.classify(region, test_shape)
+    possible_shapes = CLASSIFICATIONS[region]
+    raise "Unknown region, please provide a valid region." if possible_shapes == nil
+    if this_shape = possible_shapes.include?(test_shape)
+      "You have a(n) '#{this_shape}' arrowhead. Probably priceless."
```
Comment (fengb) [inline arrowhead/arrowhead.rb:22]: This will expand out to "You have a(n) 'true' arrowhead. Probably priceless."

Can you spot the logic error?

Comment (fengb) [inline arrowhead/arrowhead.rb:22]: This also no longer raises the error "Unknown shape value. Are you sure you know what you're talking about?"

File: robot_name/robot_name.rb [hunk 2]
```diff
@@ -7,20 +7,55 @@ class Robot
...
+    add_name_to_registry
+  end
+
+  def generate_unique_name
+    loop do 
+      # raise NameCollisionError
+      # the instructions said to make it more robust in case of a collision error
+      @name = @name_generator.generate_name
+      break if name_available?
+    end
```
Comment (fengb) [inline robot_name/robot_name.rb:24]: If I pass in a name generator that doesn't randomize, this loops infinitely. :(

File: robot_name/robot_name.rb [hunk 3]
```diff
@@ -7,20 +7,55 @@ class Robot
...
+
+  def add_name_to_registry
     @@registry << @name
   end
+
+  def name_available?
+    !(@@registry.include?(name))
+  end
+
+  def name_invalid?
```
Comment (fengb) [inline robot_name/robot_name.rb]: I usually avoid negated booleans in method names. It's easy enough to process `!name_valid?` but hard to decipher `!name_invalid?`

**Comment by fengb** [comment]
Tabbbbs!!!!!! :hankey:

File: triangle_facts/triangle_facts.rb [hunk 4]
```diff
@@ -1,45 +1,65 @@
...
+		else
+			right = "\n"
+		end
+
+		"The angles of this triangle are #{angles.join(',')}\n#{right}"
+	end
 
-		return [angleA, angleB, angleC]
+	# solving for C in a**2 + b**2 - 2abcosC = c**2
+	def calculate_single_angle(leg_1, leg_2, opposite_from_angle_of_interest)
```
Comment (fengb) [inline triangle_facts/triangle_facts.rb]: :+1: intent naming

File: triangle_facts/triangle_facts.rb [hunk 5]
```diff
@@ -1,45 +1,65 @@
 # Killer facts about triangles AWW YEAH
 class Triangle
-	attr_accessor :side1,:side2,:side3
+	attr_reader :side1, :side2, :side3
 
-	def initialize(side1,side2,side3)
-		@side1,@side2,@side3 = side1,side2,side3
+	def initialize(side1, side2, side3)
+		@side1, @side2, @side3 = side1, side2, side3
```
Comment (fengb) [inline triangle_facts/triangle_facts.rb]: Minor nitpick: I don't really like parallel assignment.

https://github.com/bbatsov/ruby-style-guide#parallel-assignment

File: triangle_facts/triangle_facts.rb [hunk 6]
```diff
@@ -1,45 +1,65 @@
...
-		unless ! (equilateral || isosceles)
-	    false
-	  else
-	    true
-	  end
+		[angleA, angleB, angleC]
+	end
+	
+	private
+	def equilateral
```
Comment (fengb) [inline triangle_facts/triangle_facts.rb]: Methods that return booleans should always _______

File: triangle_facts/triangle_facts.rb [hunk 7]
```diff
@@ -1,45 +1,65 @@
...
-		puts 'This triangle is isosceles! Also, that word is hard to type.' if isosceles 
-		puts 'This triangle is scalene and mathematically boring.' if scalene 
+	def isosceles
+		[side1,side2,side3].uniq.length == 2
+	end
 
-		angles = self.calculate_angles(side1,side2,side3)
-		puts 'The angles of this triangle are ' + angles.join(',')
+	def scalene
+		(equilateral || isosceles) ? false : true
```
Comment (fengb) [inline triangle_facts/triangle_facts.rb]: Both `equilateral` and `isosceles` are booleans so you don't need the ternary. Can you remove it while keeping the logic the same?

File: triangle_facts/triangle_facts.rb [hunk 8]
```diff
@@ -1,45 +1,65 @@
...
+		puts angle_sentences
 	end
 
-	def isosceles()
-		return [side1,side2,side3].uniq.length == 2
-	end
+	def calculate_angles(a, b, c)
+		angleA = calculate_single_angle(b, c, a)
+		angleB = calculate_single_angle(c, a, b)
+		angleC = calculate_single_angle(a, b, c)
```
Comment (fengb) [inline triangle_facts/triangle_facts.rb]: Watch out for :camel:

File: robot_name/robot_name.rb [hunk 9]
```diff
@@ -7,20 +7,55 @@ class Robot
...
 
-    raise NameCollisionError, 'There was a problem generating the robot name!' if !(name =~ /[[:alpha:]]{2}[[:digit:]]{3}/) || @@registry.include?(name)
+    add_name_to_registry
+  end
+
+  def generate_unique_name
+    loop do 
+      # raise NameCollisionError
+      # the instructions said to make it more robust in case of a collision error
+      @name = @name_generator.generate_name
```
Comment (fengb) [inline robot_name/robot_name.rb]: The previous code used `name_generator.call`. This change will break dependent code downstream.

Ideally, we should not change the logic when refactoring.

File: robot_name/robot_name.rb [hunk 10]
```diff
@@ -7,20 +7,55 @@ class Robot
...
+  def add_name_to_registry
     @@registry << @name
   end
+
+  def name_available?
+    !(@@registry.include?(name))
+  end
+
+  def name_invalid?
+    !(name =~ /[a-zA-Z]{2}[\d]{3}/)
```
Comment (fengb) [inline robot_name/robot_name.rb]: There's a `!~` operator

File: robot_name/robot_name.rb [hunk 11]
```diff
@@ -7,20 +7,55 @@ class Robot
...
-      @name = @name_generator.call
-    else
-      generate_char = -> { ('A'..'Z').to_a.sample }
-      generate_num = -> { rand(10) }
+    generate_unique_name
 
-      @name = "#{generate_char.call}#{generate_char.call}#{generate_num.call}#{generate_num.call}#{generate_num.call}"
+    # I think this error is unneccessary, but it was included in the original code
+    if name_invalid?
+      raise NameCollisionError, 'There was a problem generating the robot name!'
```
Comment (fengb) [inline robot_name/robot_name.rb]: This isn't a `NameCollisionError`. Probably needs to be renamed.

File: arrowhead/arrowhead.rb (hunk 1 above)
Comment (pmacaluso3) [inline arrowhead/arrowhead.rb:22]: This classifier is meant to determine whether the person has a true arrowhead or just a replica; working as intended. Just kidding, changed to possible_shapes[test_shape]. Unknown shape error was deleted accidentally, re-adding.

**Comment by fengb** [comment]
Please create a separate branch with only the overlord changes.

**Pull request closed without merging**